training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,846 @@
+To combine the shadow variables and write the result back into `fact`, we'll need to learn about `reduction` operations.
+
+<a id="reduce_and_scan"></a>
+## `reduce` and `scan`
+[Reductions and scans](https://chapel-lang.org/docs/language/spec/data-parallelism.html?highlight=scan#reductions-and-scans)
+cumulatively apply an operation over the elements of an array (or any iterable) in parallel.
+
+- `op scan array`
+  - scans over `array` and cumulatively applies `op` to every element
+  - returns an array
+  - `+ scan a` behaves like [`np.cumsum(a)`](https://numpy.org/doc/stable/reference/generated/numpy.cumsum.html) in numpy
+- `op reduce array`
+  - reduces the result of a scan to a single summary value
+  - returns a single value
+  - `+ reduce a` behaves like [`sum(a)`](https://docs.python.org/3/library/functions.html#sum) in python
+
+<a id="ex5"></a>
+#### Example 5: `reduce` and `scan`
+```Chapel
+var a: [0..#5] int = [1, 2, 3, 4, 5];
+writeln("a = ", a, "\n");
+writeln("+ scan a = ", + scan a);
+writeln("+ reduce a = ", + reduce a, "\n");
+writeln("* scan a = ", * scan a);
+writeln("* reduce a = ", * reduce a, "\n");
+writeln("min scan a = ", min scan a);
+writeln("min reduce a = ", min reduce a, "\n");
+writeln("max scan a = ", max scan a);
+writeln("max reduce a = ", max reduce a);
+```
+```console
+a = 1 2 3 4 5
+
++ scan a = 1 3 6 10 15
++ reduce a = 15
+
+* scan a = 1 2 6 24 120
+* reduce a = 120
+
+min scan a = 1 1 1 1 1
+min reduce a = 1
+
+max scan a = 1 2 3 4 5
+max reduce a = 5
+```
+Notice `op reduce array` is often the last element of `op scan array`. 
+
+Reductions and scans are defined for [many operations](https://chapel-lang.org/docs/language/spec/data-parallelism.html?highlight=scan#reduction-expressions)
+, and you can even [write your own](https://chapel-lang.org/docs/technotes/reduceIntents.html#user-defined-reduction-example)!
+
+<a id="forall"></a>
+## `forall` Loops
+<a id="ex6"></a>
+#### Example 6: Factorial with must-parallel `forall` and Reduction
+Back to parallel factorial, let's use `reduce` to combine the results of each task. 
+```Chapel
+proc factorial(n: int) {
+  var fact: int = 1;
+  forall i in 1..n with (* reduce fact) {
+    fact = i;
+  }
+  return fact;
+}
+writeln(factorial(5));
+```
+```console
+120
+```
+That's more like it!
+Every iteration assigned its shadow variable of `fact` to `i`, so a `* reduce` of the shadow variables gives the product of all `i`.
+
+This is an example of a must-parallel `forall`.
+
+There are [2 types of `forall`](https://chapel-lang.org/docs/primers/forallLoops.html#forall-loops) loops:
+- must-parallel `forall`
+  - Is written using the `forall` keyword i.e. `forall i in D`
+  - Requires a parallel iterator
+- may-parallel `forall`
+  - Is written using bracket notation i.e. `[i in D]`
+  - Will use a parallel iterator if present, otherwise will iterate serially
+
+<a id="ex7"></a>
+#### Example 7: may-parallel `forall`
```

Comment:
Would it be worth also demonstrating when a parallel iterator does not exist? I am not entirely sure it will add much value

Reply:
Yeah I don't think it's worth it. The [example ](https://chapel-lang.org/docs/primers/forallLoops.html#may-parallel-forall-statement) in the chapel docs (copied below) on this seems unlikely to be encountered by a new arkouda dev

```Chapel
iter onlySerial(m: int) {
  for j in 1..m do
    yield j;
}

[i in onlySerial(n)] {
  writeln("in onlySerial iteration #", i);
}
```
```Chapel
forall i in onlySerial(n) { // error: a parallel iterator is not found
  writeln("in iteration #", i);
}
```